medikanren/lru.rkt

```diff
@@ -0,0 +1,234 @@
+#lang racket
+(provide
+ make-lru
+ lru-ref
+ )
+(require chk)
+
+
+(define (assert k st)
+  (if (not k)
+      (raise (format "assertion failure: ~a" st))
+      #f))
+
+(struct payload (k v))
+
+(struct lrun (
+              (older #:mutable)
+              payload
+              (newer #:mutable)))
+
+(struct lru (
+             behind-ref
+             hash
+             (lrun-oldest #:mutable)
+             (lrun-newest #:mutable)
+             (num-entries #:mutable)
+             num-entries-max))
+
+(define (make-lru behind-ref #:num-entries-max (num-entries-max 1000))
+  (lru
+   behind-ref
+   (make-hash)
+   #f
+   #f
+   0
+   num-entries-max))
+
+;;; Add a newest entry to the lru.
+(define (lru-put-newest ths payload1)
+  (let* (
+         (lrun1 (lru-lrun-newest ths))
+         (k (payload-k payload1))
+         (lrun0 (lrun lrun1 payload1 #f))
+         )
+    ; connect lrun fields
+    ; lrun0.older is already connected
+    (set-lrun-older! lrun0 lrun1) 
+    (if lrun1
+        (set-lrun-newer! lrun1 lrun0)
+        #f)
+    ; connect lru fields
+    (set-lru-lrun-newest! ths lrun0)
+    (if (not (lru-lrun-oldest ths)) ; are we brand new?
+        (set-lru-lrun-oldest! ths lrun0)
+        #f)
+    (set-lru-num-entries! ths (+ (lru-num-entries ths) 1))
+    (hash-set! (lru-hash ths) k lrun0)))
+
+(define (lru-remove ths lrun1)
+  (let* (
+         (payload1 (lrun-payload lrun1))
+         (k (payload-k payload1))
+         (lrun0 (lrun-older lrun1))
+         (lrun2 (lrun-newer lrun1))
+         )
+    (if lrun0
+        (set-lrun-newer! lrun0 lrun2)
+        (begin
+          ; we are removing the oldest
+          (set-lru-lrun-oldest! ths lrun2)
+          (set-lrun-older! lrun2 #f)))
```

Comment:
`lrun2` may be `#f` here.  Line 73 already takes care of this though, so it should be fine to delete this line.